pylint/extensions/typing.py

```diff
@@ -219,6 +224,66 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
         if self._should_check_callable and node.attrname == "Callable":
             self._check_broken_callable(node)
 
+    @only_required_for_messages("redundant-union-assign-typehint")
+    def visit_annassign(self, node: nodes.AnnAssign) -> None:
+        type_annotations = list(node.annotation.get_children())
+        if not type_annotations:
+            return
+        if (
+            isinstance(type_annotations[0], nodes.Name)
+            and type_annotations[0].name in UNION_NAMES
+        ):
+            if isinstance(type_annotations[-1], nodes.Tuple):
+                types = list(type_annotations[-1].get_children())
+            elif type_annotations[0].name == "Optional":
+                if (
+                    isinstance(type_annotations[-1], nodes.Const)
+                    and type_annotations[-1].pytype() == "builtins.NoneType"
+                ):
+                    self.add_message(
+                        "redundant-union-assign-typehint",
+                        node=node,
+                        args="None",
+                        confidence=HIGH,
+                    )
+                return
+        elif self._should_check_alternative_union_syntax and isinstance(
+            node.annotation, nodes.BinOp
+        ):
+            types = self.parse_binops_typehints(
+                node.annotation,
+            )
+        else:
+            return
+        redundant_typehints: list[str] = []
+        for idx, typehint in enumerate(types):
+            typehint_str = typehint.as_string()
+            if (
+                (idx < (len(types) - 1))
+                and (any(typehint_str == x.as_string() for x in types[idx + 1 :]))
+                and typehint_str not in redundant_typehints
+            ):
+                redundant_typehints.append(typehint_str)
```

Comment:
This is quite inefficient. `x.as_string()` is basically called `n * n` times while not being cached. Additionally, the `typehint` on which the error would be emitted is the first occurrence and not the second, third, ...

A better idea, try using a dict with the `typehint_str` as key and the position as value. (A set might also work if we omit the position.)
1. Compute `typehint_str`
2. Check if the str is in the dict / set
3. If exist emit error
4. Else add item to dict / set

Reply:
@cdce8p  - beautiful algorithm mate! using a set is very clever. Thanks!